src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -306,6 +307,20 @@ struct TreeVisitor<'tree> {
     perms: &'tree mut UniValMap<LocationState>,
 }
 
+pub enum AccessRangeAndKind<'a, 'tcx> {
+    Regular { range: AllocRange, kind: AccessKind, cause: diagnostics::AccessCause },
+    MagicProtectorEnd(Option<&'a VClockAlloc>, &'a MiriMachine<'tcx>),
+}
```

Comment:
Could you factor this new enum into a separate PR? That seems worth having independently of this -- and it's always easier if commits without functional changes are split from those with functional changes.